### PR TITLE
Filter unknown enums when removing unknown parameters

### LINKED ISSUE #2857: Core rejects RAI if unknown enums are present
### Bug Report
Core by default rejects any message with `INVALID_DATA` if it contains any unknown enum values, but this should not be done for the RegisterAppInterface request, since this message is sent before version negotiation occurs

##### Reproduction Steps
1. Send a RegisterAppInterface request with a new AppHMIType (ex. "SHOPPING") and an existing AppHMIType (ex. "MEDIA")

##### Expected Behavior
Core sends a `SUCCESS` response with the negotiated version, which can be used by the application to determine that the new AppHMIType is not supported by the module

##### Observed Behavior
Core responds with `INVALID_DATA`

##### OS & Version Information
* OS/Version: Ubuntu 18.04
* SDL Core Version: 5.0.0
* Testing Against: iOS test app

### PR DESCRIPTION
**Pull request by jacobkeeler**
Fixes #2857
This PR is **ready** for review.

### Risk
This PR makes **minor** API changes.

### Testing Plan
Follow issue reproduction steps in #2857 

### Summary
Change schema logic to filter out unknown enum values when removing unknown parameter

##### Bug Fixes
* Fixes issue where RAI was rejected if it contained new enums before version negotiation was completed.

### CLA
- [x] I have signed [the CLA](https://docs.google.com/forms/d/e/1FAIpQLSdsgJY33VByaX482zHzi-xUm49JNnmuJOyAM6uegPQ2LXYVfA/viewform)

### CODE REVIEW

File: src/components/application_manager/src/commands/command_request_impl.cc [hunk 1]
```diff
@@ -298,7 +298,7 @@ void CommandRequestImpl::SendResponse(
     response[strings::msg_params] = *response_params;
   }
 
-  if (info) {
+  if (info && !info.empty()) {
```
Comment (JackLivio) [inline src/components/application_manager/src/commands/command_request_impl.cc:301]: Does not compile since its a char array, need to check if first character is '\0'

```suggestion
  if (info && *info != '\0') {
```

File: src/components/application_manager/src/commands/command_request_impl.cc [hunk 2]
```diff
@@ -318,7 +318,17 @@ void CommandRequestImpl::SendResponse(
   }
 
   response[strings::msg_params][strings::success] = success;
-  response[strings::msg_params][strings::result_code] = result_code;
+  if ((result_code == mobile_apis::Result::SUCCESS ||
+       result_code == mobile_apis::Result::WARNINGS) &&
+      !warning_info().empty()) {
+    response[strings::msg_params][strings::info] =
+        (info && !info.empty()) ? std::string(info) + "\n" + warning_info()
```
Comment (JackLivio) [inline src/components/application_manager/src/commands/command_request_impl.cc:325]:
```suggestion
        (info && *info != '\0') ? std::string(info) + "\n" + warning_info()
```

File: src/components/smart_objects/include/smart_objects/enum_schema_item.h [hunk 3]
```diff
@@ -314,6 +317,21 @@ const ElementSignature TEnumSchemaItem<EnumType>::getSignature(
...
 
+template <typename EnumType>
+bool TEnumSchemaItem<EnumType>::filterInvalidEnums(
+    SmartObject& Object,
+    const utils::SemanticVersion& MessageVersion,
+    rpc::ValidationReport* report__) {
+  rpc::ValidationReport dummy_report("");
+  if (validate(Object, &dummy_report, MessageVersion, false) != errors::OK) {
+    std::string validation_info =
+        "Filtered invalid value - " + Object.asString();
```
Comment (JackLivio) [inline src/components/smart_objects/include/smart_objects/enum_schema_item.h:328]: Not sure if this is the validation info passed back to mobile but if it is, then it needs to follow the format mentioned in the proposal:

> Invalid enums were removed: <param_1>:<enum_value_1>,<enum_value_2>;<param_2>:<enum_value_3> ...

File: src/components/smart_objects/include/smart_objects/enum_schema_item.h [hunk 4]
```diff
@@ -128,15 +128,17 @@ class TEnumSchemaItem : public CDefaultSchemaItem<EnumType> {
       const std::vector<ElementSignature>& signatures,
       const utils::SemanticVersion& MessageVersion);
 
+  bool filterInvalidEnums(SmartObject& Object,
```
Comment (JackLivio) [inline src/components/smart_objects/include/smart_objects/enum_schema_item.h:131]: Please add method description of params and return value

File: src/components/smart_objects/include/smart_objects/schema_item.h [hunk 5]
```diff
@@ -91,13 +101,27 @@ class ISchemaItem {
...
 
+  /**
+   * @brief Filter invalid enum values
+   *
+   * @param Object Object to check for invalid enum values
+   * @param MessageVersion the version of the schema to use for validation
+   * @param report__ object for reporting enums which were removed during the
+   * process
+   *
+   * @return true if the value being checked should be filtered, false otherwise
```
Comment (JackLivio) [inline src/components/smart_objects/include/smart_objects/schema_item.h:112]: The return value is a bit confusing. It says if true it should be filtered, but isnt the enum implementation of this method doing the filtering?

File: src/components/smart_objects/src/array_schema_item.cc [hunk 6]
```diff
@@ -90,6 +92,31 @@ errors::eType CArraySchemaItem::validate(
...
+      strVal << index;
+      if (mElementSchemaItem->filterInvalidEnums(
+              *it, MessageVersion, &report__->ReportSubobject(strVal.str()))) {
+        it = array->erase(it);
+      } else {
+        ++it;
+      }
+      index++;
+    }
+    if (array->empty()) {
```
Comment (JackLivio) [inline src/components/smart_objects/src/array_schema_item.cc:113]: Should this method return true if there was at least enum item removed?

File: src/components/smart_objects/src/smart_schema.cc [hunk 7]
```diff
@@ -59,8 +59,16 @@ ISchemaItemPtr CSmartSchema::getSchemaItem() {
...
-                               const utils::SemanticVersion& MessageVersion) {
+                               const utils::SemanticVersion& MessageVersion,
+                               rpc::ValidationReport* report__) {
   mSchemaItem->applySchema(Object, remove_unknown_parameters, MessageVersion);
+  if (remove_unknown_parameters) {
+    rpc::ValidationReport dummy_report("");
+    if (!report__) {
+      report__ = &dummy_report;
+    }
+    mSchemaItem->filterInvalidEnums(Object, MessageVersion, report__);
```
Comment (JackLivio) [inline src/components/smart_objects/src/smart_schema.cc:70]: Should the return value be checked here?

Comment (JackLivio) [inline src/components/smart_objects/src/smart_schema.cc:70]: Its a little confusing that the remove invalid enums was put in applySchema and not in `RemoveUnkownParams`. Any reason why it was split like this?

Comment (JackLivio) [inline src/components/smart_objects/src/smart_schema.cc:70]: Should the call to filter unknown enums be moved to RemoveUnknownParams? The code is all over the place and its pretty hard to follow whats happening where.

File: src/components/application_manager/src/rpc_service_impl.cc [hunk 8]
```diff
@@ -245,6 +246,12 @@ bool RPCServiceImpl::ManageMobileCommand(
           app->hmi_level(mobile_apis::PredefinedWindows::DEFAULT_WINDOW);
     }
 
+    commands::CommandRequestImpl* request =
+        dynamic_cast<commands::CommandRequestImpl*>(command.get());
+    if (request != NULL) {
+      request->set_warning_info(warning_info);
```
Comment (LuxoftAKutsan) [inline src/components/application_manager/src/rpc_service_impl.cc:252]: Can we move set_warning_info to base class?

File: src/components/formatters/include/formatters/CSmartFactory.h [hunk 9]
```diff
@@ -277,7 +279,8 @@ template <class FunctionIdEnum, class MessageTypeEnum, class StructIdEnum>
 bool CSmartFactory<FunctionIdEnum, MessageTypeEnum, StructIdEnum>::attachSchema(
     ns_smart_device_link::ns_smart_objects::SmartObject& object,
     const bool remove_unknown_parameters,
-    const utils::SemanticVersion& MessageVersion) {
+    const utils::SemanticVersion& MessageVersion,
+    rpc::ValidationReport* report__) {
```
Comment (LuxoftAKutsan) [inline src/components/formatters/include/formatters/CSmartFactory.h:283]: Do we need 2 underscores after parameter name?

File: src/components/include/application_manager/rpc_service.h [hunk 10]
```diff
@@ -54,10 +54,15 @@ class RPCService {
    * @brief ManageMobileCommand convert message to mobile command and execute it
    * @param message pointer to received message
    * @param origin origin of command
+   * @param warning_info warning message to send on a successful response. Only
+   * applies to requests from mobile.
    * @return true if command is executed, otherwise return false
    */
-  virtual bool ManageMobileCommand(const commands::MessageSharedPtr message,
-                                   commands::Command::CommandSource source) = 0;
+  virtual bool ManageMobileCommand(
```
Comment (LuxoftAKutsan) [inline src/components/include/application_manager/rpc_service.h:61]: I do understand benefit of adding default argument, we will not have additional method and mock. 
But IMHO and according to [google coding stye](https://google.github.io/styleguide/cppguide.html#Default_Arguments) default arguments in virtual functions is bad practice. 
I would propose to make additional function `ManageMobileCommandExtended`/`ManageMobileCommandWarning`/`ManageMobileInfoCommand` ( naming is complicated) with warning code,  and keep ManageMobileCommand as proxy to it.

File: src/components/smart_objects/include/smart_objects/array_schema_item.h [hunk 11]
```diff
@@ -80,13 +80,17 @@ class CArraySchemaItem : public ISchemaItem {
       const utils::SemanticVersion& MessageVersion = utils::SemanticVersion(),
       const bool allow_unknown_enums = false) OVERRIDE;
 
+  bool filterInvalidEnums(SmartObject& Object,
```
Comment (LuxoftAKutsan) [inline src/components/smart_objects/include/smart_objects/array_schema_item.h:83]: Shouldn't we start from capital letter?

File: src/components/smart_objects/include/smart_objects/schema_item.h [hunk 12]
```diff
@@ -91,13 +101,27 @@ class ISchemaItem {
...
+   **/
+  virtual bool filterInvalidEnums(SmartObject& Object,
```
Comment (LuxoftAKutsan) [inline src/components/smart_objects/include/smart_objects/schema_item.h:114]: I see check of return value only in unit tests. Do we need it?

File: src/components/smart_objects/src/array_schema_item.cc [hunk 13]
```diff
@@ -90,6 +92,31 @@ errors::eType CArraySchemaItem::validate(
   return errors::OK;
 }
 
+bool CArraySchemaItem::filterInvalidEnums(
+    SmartObject& Object,
+    const utils::SemanticVersion& MessageVersion,
+    rpc::ValidationReport* report__) {
+  if (SmartType_Array == Object.getType()) {
```
Comment (LuxoftAKutsan) [inline src/components/smart_objects/src/array_schema_item.cc:99]: I would use reverse if end early exist. Just as suggestion.

File: src/components/smart_objects/src/array_schema_item.cc [hunk 14]
```diff
@@ -90,6 +92,31 @@ errors::eType CArraySchemaItem::validate(
...
+    int index = 0;
+    auto array = Object.asArray();
+    for (auto it = array->begin(); it != array->end();) {
+      std::stringstream strVal;
+      strVal << index;
```
Comment (LuxoftAKutsan) [inline src/components/smart_objects/src/array_schema_item.cc:104]: We ca use https://en.cppreference.com/w/cpp/string/basic_string/to_string

File: src/components/smart_objects/src/array_schema_item.cc [hunk 15]
```diff
@@ -90,6 +92,31 @@ errors::eType CArraySchemaItem::validate(
   return errors::OK;
 }
 
+bool CArraySchemaItem::filterInvalidEnums(
```
Comment (LuxoftAKutsan) [inline src/components/smart_objects/src/array_schema_item.cc:95]: I feel this function as rather complicated one. 
I would try probably to sacrifice performance but make it more easy to read and obvious.

File: src/components/smart_objects/src/object_schema_item.cc [hunk 16]
```diff
@@ -157,6 +158,31 @@ errors::eType CObjectSchemaItem::validate(
   return errors::OK;
 }
 
+bool CObjectSchemaItem::filterInvalidEnums(
+    SmartObject& Object,
+    const utils::SemanticVersion& MessageVersion,
+    rpc::ValidationReport* report__) {
+  for (const auto& key : Object.enumerate()) {
```
Comment (LuxoftAKutsan) [inline src/components/smart_objects/src/object_schema_item.cc:165]: Why don't you check  `Object.getType` ?

File: src/components/smart_objects/include/smart_objects/schema_item.h (hunk 12 above)
Comment (jacobkeeler) [reply src/components/smart_objects/include/smart_objects/schema_item.h:114]: Yes, it is used recursively within the function. If a particular element returns true, that means that it should be removed from the containing structure. This return value is not used at the root level because that would mean that the full structure would be destroyed.

File: src/components/smart_objects/src/array_schema_item.cc (hunk 13 above)
Comment (jacobkeeler) [reply src/components/smart_objects/src/array_schema_item.cc:99]: Are you just suggesting the operands be swapped? I'm a bit confused by this comment

File: src/components/smart_objects/src/array_schema_item.cc (hunk 6 above)
Comment (jacobkeeler) [reply src/components/smart_objects/src/array_schema_item.cc:113]: No, that would mean that the array should be removed, which doesn't make sense in that case

Comment (jacobkeeler) [reply src/components/smart_objects/src/array_schema_item.cc:113]: That said, it would probably be better to return if the list is shorter than the minimum length, rather than explicitly checking for an empty list

File: src/components/smart_objects/src/object_schema_item.cc (hunk 16 above)
Comment (jacobkeeler) [reply src/components/smart_objects/src/object_schema_item.cc:165]: Ah, I added the `GetType` function for use with my original (and simpler) implementation of this, it didn't turn out to be necessary in the final implementation, so I can remove that function.

File: src/components/smart_objects/src/smart_schema.cc (hunk 7 above)
Comment (jacobkeeler) [reply src/components/smart_objects/src/smart_schema.cc:70]: I could leave a log here if that works, not much else we could do with it though in this case.

File: src/components/smart_objects/src/array_schema_item.cc (hunk 15 above)
Comment (jacobkeeler) [reply src/components/smart_objects/src/array_schema_item.cc:95]: Perhaps I might be able to add some extra comments for clarity. I don't think there's really a way to make it simpler though

File: src/components/smart_objects/include/smart_objects/enum_schema_item.h (hunk 4 above)
Comment (jacobkeeler) [reply src/components/smart_objects/include/smart_objects/enum_schema_item.h:131]: It's in the `schema_item.h` header, we should only have it in one place

File: src/components/smart_objects/include/smart_objects/array_schema_item.h (hunk 11 above)
Comment (jacobkeeler) [reply src/components/smart_objects/include/smart_objects/array_schema_item.h:83]: Perhaps by normal case rules, I was following the standards of this class in this case (ex. `validate`, `applySchema`, etc.)

File: src/components/formatters/include/formatters/CSmartFactory.h (hunk 9 above)
Comment (jacobkeeler) [reply src/components/formatters/include/formatters/CSmartFactory.h:283]: Ah, just following previous examples, I can remove it

File: src/components/application_manager/src/rpc_service_impl.cc (hunk 8 above)
Comment (jacobkeeler) [reply src/components/application_manager/src/rpc_service_impl.cc:252]: I can't see it being useful for any other type of message, but I do agree that casting it down is a bit ugly. I can move it to make things a bit more clean.

**Comment by jacobkeeler** [comment]
@LuxoftAKutsan ah yes, this was originally just to fix the one issue, but after discovering the proposal I changed it to match the proposal requirements. I added the proposal to the PR description accordingly.

File: src/components/smart_objects/src/smart_schema.cc (hunk 7 above)
Comment (jacobkeeler) [reply src/components/smart_objects/src/smart_schema.cc:70]: Ah, sadly the logger is not available in the smart objects component, so there really isn't anything to do here.

File: src/components/formatters/include/formatters/CSmartFactory.h (hunk 9 above)
Comment (jacobkeeler) [reply src/components/formatters/include/formatters/CSmartFactory.h:283]: Fixed in https://github.com/smartdevicelink/sdl_core/pull/3385/commits/bee65b5d8fe27964ed09db991e9ecd8c3b7d8d2c

File: src/components/application_manager/src/rpc_service_impl.cc (hunk 8 above)
Comment (jacobkeeler) [reply src/components/application_manager/src/rpc_service_impl.cc]: Fixed in https://github.com/smartdevicelink/sdl_core/pull/3385/commits/6840997e72ef482f670c747b72ac78ecc8d29fb3

File: src/components/smart_objects/src/array_schema_item.cc (hunk 15 above)
Comment (jacobkeeler) [reply src/components/smart_objects/src/array_schema_item.cc:93]: Added comments in https://github.com/smartdevicelink/sdl_core/pull/3385/commits/147b7bca79fe459cdceeeee6d7b8a74133890571

File: src/components/smart_objects/src/object_schema_item.cc (hunk 16 above)
Comment (jacobkeeler) [reply src/components/smart_objects/src/object_schema_item.cc:166]: I ended up using it in this class to filter out single invalid enums from the object

File: src/components/smart_objects/src/array_schema_item.cc (hunk 14 above)
Comment (jacobkeeler) [reply src/components/smart_objects/src/array_schema_item.cc]: Fixed in https://github.com/smartdevicelink/sdl_core/pull/3385/commits/bee65b5d8fe27964ed09db991e9ecd8c3b7d8d2c

File: src/components/smart_objects/src/array_schema_item.cc (hunk 6 above)
Comment (jacobkeeler) [reply src/components/smart_objects/src/array_schema_item.cc]: Ah, I think I see your point, it should return true ONLY if at least one item is removed. Fixed in https://github.com/smartdevicelink/sdl_core/pull/3385/commits/bee65b5d8fe27964ed09db991e9ecd8c3b7d8d2c

File: src/components/include/application_manager/rpc_service.h (hunk 10 above)
Comment (jacobkeeler) [reply src/components/include/application_manager/rpc_service.h]: Changed to overload in https://github.com/smartdevicelink/sdl_core/pull/3385/commits/446fe5c5c95865f6d7d088a088cbd5d655de2851

**Comment by jacobkeeler** [comment]
@LuxoftAKutsan @JackLivio addressed comments

File: src/components/smart_objects/src/array_schema_item.cc (hunk 13 above)
Comment (LuxoftAKutsan) [reply src/components/smart_objects/src/array_schema_item.cc:98]: Consider following approach : 
```cpp
if (SmartType_Array != Object.getType()) { 
  return false;
  // This block may be removed within removing Object.getType(
}

auto array = Object.asArray();
auto elements_to_remove = std::find_if(array->begin(), array->end(), []() {
  mElementSchemaItem->filterInvalidEnums(...)
});
for (element: elements_to_remove) {
  array.eraze(element);
}
// min size staff using elements_to_remove.size()
}
```
I didn't compile it, so there could be issues with syntax. 
 
I see this benefits : 
 - No nested structures 
 - No iterator crutches.

I agree that it probably less efficient then previous one. But we do not expect huge amount of elements in `Object`, do we?

File: src/components/smart_objects/src/object_schema_item.cc [hunk 17]
```diff
@@ -157,6 +158,39 @@ errors::eType CObjectSchemaItem::validate(
   return errors::OK;
 }
 
+bool CObjectSchemaItem::filterInvalidEnums(
+    SmartObject& Object,
+    const utils::SemanticVersion& MessageVersion,
+    rpc::ValidationReport* report) {
+  bool valid = true;
+  for (const auto& key : Object.enumerate()) {
+    std::map<std::string, SMember>::const_iterator members_it =
```
Comment (LuxoftAKutsan) [inline src/components/smart_objects/src/object_schema_item.cc:167]: good place for auto

File: src/components/smart_objects/src/object_schema_item.cc [hunk 18]
```diff
@@ -157,6 +158,39 @@ errors::eType CObjectSchemaItem::validate(
   return errors::OK;
 }
 
+bool CObjectSchemaItem::filterInvalidEnums(
```
Comment (LuxoftAKutsan) [inline src/components/smart_objects/src/object_schema_item.cc:161]: Consider following approach. 
```cpp


bool CObjectSchemaItem::filterInvalidEnums(
    SmartObject& Object,
    const utils::SemanticVersion& MessageVersion,
    rpc::ValidationReport* report) {
  
  auto filter_invalid_enums = [this, &MessageVersion, report](member) {
      const SMember* member =
          GetCorrectMember(members_it->second, MessageVersion);
      rreturn member->mSchemaItem->filterInvalidEnums(
              Object[key], MessageVersion, &report->ReportSubobject(key));
  };

  auto should_be_erazed = [](member) -> bool {
        // The member is safe to filter if it is non-mandatory
        // Object is no longer valid if the member is mandatory.
        // To maintain the general structure of the message, only leaf nodes
        // (individual enum values) should be filtered in this case.
      return member->mSchemaItem->GetType() == TYPE_ENUM && !member->mIsMandatory;
  };

  bool valid = true;
  for (const auto& key : Object.enumerate()) {
    auto members_it = mMembers.find(key);
    DCHECK_OR_RETURN(mMembers.end() != members_it, false);  //  object should match schema
    if (filter_invalid_enums(*members_it)) {
        if (member->mIsMandatory) {
          valid = false;
        }
        if (should_be_erazed(member)) {
            Object.erase(key); 
        }
    }
  }
  return !valid;
}
```

The same as above, I didn't check it and it not so performant. but I feel it as more readable.

File: src/components/smart_objects/test/ArraySchemaItem_test.cc [hunk 19]
```diff
@@ -345,6 +349,92 @@ TEST(test_map_validate, test_ArraySchemaItemTest) {
...
+
+  rpc::ValidationReport report("RPC");
+  EXPECT_TRUE(item->filterInvalidEnums(obj, utils::SemanticVersion(), &report));
+  EXPECT_NE(std::string(""), rpc::PrettyFormat(report));
+
+  EXPECT_EQ((size_t)0, obj.length());
+
+  report = rpc::ValidationReport("RPC");
+  EXPECT_EQ(errors::OUT_OF_RANGE, item->validate(obj, &report));
+  EXPECT_NE(std::string(""), rpc::PrettyFormat(report));
```
Comment (LuxoftAKutsan) [inline src/components/smart_objects/test/ArraySchemaItem_test.cc:410]: Why do we expect empty report?

File: src/components/smart_objects/test/ArraySchemaItem_test.cc [hunk 20]
```diff
@@ -345,6 +349,92 @@ TEST(test_map_validate, test_ArraySchemaItemTest) {
...
+}
```
Comment (LuxoftAKutsan) [inline src/components/smart_objects/test/ArraySchemaItem_test.cc:411]: Would be good also to cover with UT : 
 - filtering of mandatory\not mandatory
 - filtering of nested structures with mandatory\not mandatory
 - warning message

File: src/components/smart_objects/test/ArraySchemaItem_test.cc (hunk 19 above)
Comment (jacobkeeler) [reply src/components/smart_objects/test/ArraySchemaItem_test.cc:410]: We expect a non-empty report (`EXPECT_NE`). I don't think we should enforce a strict format, just verify that some warning was generated

File: src/components/smart_objects/include/smart_objects/array_schema_item.h [hunk 21]
```diff
@@ -113,6 +117,8 @@ class CArraySchemaItem : public ISchemaItem {
   void BuildObjectBySchema(const SmartObject& pattern_object,
                            SmartObject& result_object) OVERRIDE;
 
+  TypeID GetType() OVERRIDE;
```
[the remark this answers is not in the document]
Comment (jacobkeeler) [reply src/components/smart_objects/include/smart_objects/array_schema_item.h:120]: Per my later comment (https://github.com/smartdevicelink/sdl_core/pull/3385#discussion_r436099463), I found that it was needed in the latest changes to `object_schema_item.cc`

File: src/components/smart_objects/include/smart_objects/enum_schema_item.h [hunk 22]
```diff
@@ -314,10 +317,25 @@ const ElementSignature TEnumSchemaItem<EnumType>::getSignature(
   return ret;
 }
 
+template <typename EnumType>
+bool TEnumSchemaItem<EnumType>::filterInvalidEnums(
+    SmartObject& Object,
+    const utils::SemanticVersion& MessageVersion,
+    rpc::ValidationReport* report) {
+  rpc::ValidationReport dummy_report("");
+  if (validate(Object, &dummy_report, MessageVersion, false) != errors::OK) {
```
[the remark this answers is not in the document]
Comment (jacobkeeler) [reply src/components/smart_objects/include/smart_objects/enum_schema_item.h:326]: Since `false` is the default value for this parameter, I could just remove it from this call

File: src/components/smart_objects/src/object_schema_item.cc (hunk 18 above)
Comment (jacobkeeler) [reply src/components/smart_objects/src/object_schema_item.cc:161]: Seems a lambda would be overkill for "should_be_erazed" when a local variable would work fine. IMO unnecessary lambdas make things less clear overall. 

I'll see what I can do to clean it up.

File: src/components/smart_objects/test/ArraySchemaItem_test.cc (hunk 20 above)
Comment (jacobkeeler) [reply src/components/smart_objects/test/ArraySchemaItem_test.cc:417]: Added tests: https://github.com/smartdevicelink/sdl_core/pull/3385/commits/e547238c73dc223287a8ac46cfbe6a6e92e3c18c

File: src/components/smart_objects/src/object_schema_item.cc (hunk 17 above)
Comment (jacobkeeler) [reply src/components/smart_objects/src/object_schema_item.cc]: Fixed in https://github.com/smartdevicelink/sdl_core/pull/3385/commits/e547238c73dc223287a8ac46cfbe6a6e92e3c18c

File: src/components/smart_objects/include/smart_objects/enum_schema_item.h (hunk 22 above)
[the remark this answers is not in the document]
Comment (jacobkeeler) [reply src/components/smart_objects/include/smart_objects/enum_schema_item.h]: Fixed in https://github.com/smartdevicelink/sdl_core/pull/3385/commits/e547238c73dc223287a8ac46cfbe6a6e92e3c18c

File: src/components/application_manager/include/application_manager/rpc_handler_impl.h [hunk 23]
```diff
@@ -176,6 +176,7 @@ class RPCHandlerImpl : public RPCHandler,
 
   bool ConvertMessageToSO(const Message& message,
                           smart_objects::SmartObject& output,
+                          std::string& warning_info,
```
[the remark this answers is not in the document]
Comment (jacobkeeler) [reply src/components/application_manager/include/application_manager/rpc_handler_impl.h]: Fixed in https://github.com/smartdevicelink/sdl_core/pull/3385/commits/e547238c73dc223287a8ac46cfbe6a6e92e3c18c

File: src/components/smart_objects/src/object_schema_item.cc (hunk 18 above)
Comment (jacobkeeler) [reply src/components/smart_objects/src/object_schema_item.cc:161]: Refactored in https://github.com/smartdevicelink/sdl_core/pull/3385/commits/b4452abf1c2f607f6dc68cb96889cdaff8363dde

**Comment by jacobkeeler** [comment]
@LuxoftAKutsan addressed new comments

**Comment by LuxoftAKutsan** [comment]
I see that proposal implementation merged to develop. 
Is there ant of following components: 
 - Regression testing result
 - Manual test plan. 
 - Don't we expect any changes in automated test coverage that may trigger changes in sdl_core?
